EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -251,23 +249,58 @@ public static int Compute_PSM_PEP(List<PeptideSpectralMatch> psms, List<int> psm
             return ambiguousPeptidesResolved;
         }
 
-        public static List<int>[] Get_PSM_Group_Indices(int psmsCount, int numGroups)
+        //we add the indexes of the targets and decoys to the groups separately in the hope that we'll get at least one target and one decoy in each group.
+        //then training can possibly be more successful.
+        public static List<int>[] Get_PSM_Group_Indices(List<PeptideSpectralMatch> psms, int numGroups)
         {
             List<int>[] groupsOfIndicies = new List<int>[numGroups];
             for (int i = 0; i < numGroups; i++)
             {
                 groupsOfIndicies[i] = new List<int>();
             }
 
+            List<int> targetPsmIndexes = new List<int>();
+            List<int> decoyPsmIndexes = new List<int>();
+
+            for (int i = 0; i < psms.Count; i++)
+            {
+                if (psms[i].IsDecoy)
+                {
+                    decoyPsmIndexes.Add(i);
+                }
+                else
+                {
+                    targetPsmIndexes.Add(i);
+                }
+            }
+
             int myIndex = 0;
-            while (myIndex < psmsCount)
+            if (decoyPsmIndexes.Count > 0)
```

Comment:
This check is unnecessary; the while loop will exit if this is false

Reply:
Unclear how PEP trains if there are no decoys...

Reply:
I removed the unnecessary checks. 

PEP will not train if any of the four sets is missing both targets and decoys. The PEP columns will be empty. 

![image](https://user-images.githubusercontent.com/16841846/103557797-58a2b680-4e79-11eb-9447-399e7bbdecfd.png)